bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/test/java/com/google/cloud/bigtable/mirroring/hbase1_x/bufferedmutator/TestConcurrentMirroringBufferedMutator.java

```diff
@@ -0,0 +1,363 @@
+        .asyncRequestResource(any(RequestResourcesDescription.class));
+    verify(common.resourceReservation, never()).release();
+  }
+
+  @Test
+  public void testBufferedMutatorFlush() throws IOException {
+    BufferedMutator bm = getBufferedMutator((long) (common.mutationSize * 3.5));
+
+    bm.mutate(common.mutation1);
+    bm.mutate(common.mutation1);
+    bm.mutate(common.mutation1);
+    bm.flush();
+    executorServiceRule.waitForExecutor();
+    verify(common.primaryBufferedMutator, times(3)).mutate(singletonMutation1);
+    verify(common.secondaryBufferedMutator, times(3)).mutate(singletonMutation1);
+    verify(common.primaryBufferedMutator, times(1)).flush();
+    verify(common.secondaryBufferedMutator, times(1)).flush();
+    verify(common.flowController, never())
+        .asyncRequestResource(any(RequestResourcesDescription.class));
+    verify(common.resourceReservation, never()).release();
+  }
+
+  @Test
+  public void testCloseFlushesWrites() throws IOException {
+    BufferedMutator bm = getBufferedMutator((long) (common.mutationSize * 3.5));
+
+    bm.mutate(common.mutation1);
+    bm.mutate(common.mutation1);
+    bm.mutate(common.mutation1);
+    bm.close();
+    verify(common.primaryBufferedMutator, times(3)).mutate(singletonMutation1);
+    verify(common.secondaryBufferedMutator, times(3)).mutate(singletonMutation1);
+    verify(common.primaryBufferedMutator, times(1)).flush();
+    verify(common.secondaryBufferedMutator, times(1)).flush();
+    verify(common.flowController, never())
+        .asyncRequestResource(any(RequestResourcesDescription.class));
+    verify(common.resourceReservation, never()).release();
+  }
+
+  @Test
+  public void testCloseIsIdempotent() throws IOException {
+    BufferedMutator bm = getBufferedMutator((long) (common.mutationSize * 3.5));
+
+    bm.mutate(common.mutation1);
+    bm.mutate(common.mutation1);
+    bm.mutate(common.mutation1);
+    bm.close();
+    bm.close();
+    verify(common.secondaryBufferedMutator, times(1)).flush();
+    verify(common.flowController, never())
+        .asyncRequestResource(any(RequestResourcesDescription.class));
+    verify(common.resourceReservation, never()).release();
+  }
+
+  @Test
+  public void testBlockedPrimaryFlushesDoNotPreventSecondaryAsyncFlushes() throws IOException {
+    final SettableFuture<Void> startPrimaryFlush = SettableFuture.create();
+    final SettableFuture<Void> startSecondaryFlush = SettableFuture.create();
+
+    blockMethodCall(common.primaryBufferedMutator, startPrimaryFlush).flush();
+    blockMethodCall(common.secondaryBufferedMutator, startSecondaryFlush).flush();
+
+    BufferedMutator bm = getBufferedMutator((long) (common.mutationSize * 0.5));
+
+    bm.mutate(common.mutation1);
+    bm.mutate(common.mutation2);
+    startPrimaryFlush.set(null);
+    executorServiceRule.waitForExecutor();
+
+    verify(common.primaryBufferedMutator, times(2)).flush();
+    verify(common.secondaryBufferedMutator, times(1)).flush();
```

Comment:
why secondary is executed only once? Also how does this test validates that primary is blocked and secondary is not?

Reply:
I've added some comments to explain what is going on. I've also modified this test to make some of the things more explicit, especially it turned out that in this specific test `waitForExecutor()` call was in fact equivalent to `sleep(3 seconds)` - I've modified it to make waiting for things to happen (or not) explicit.